File: 2017/09/09.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	f, err := os.Open("input")
	if err != nil {
		panic(err)
	}
	defer f.Close()

	s := bufio.NewScanner(f)
	s.Split(bufio.ScanRunes)

	groupLevel := 0
	groupSum := 0
	garbageCount := 0
	var garbage bool
	for s.Scan() {
		char := s.Text()
		//fmt.Println(char)
		switch char {
		case "{":
			if !garbage {
				groupLevel++
				groupSum += groupLevel
			} else {
				garbageCount++
			}
		case "}":
			if !garbage {
				groupLevel--
			} else {
				garbageCount++
			}
		case "<":
			if !garbage {
				garbage = true
			} else {
				garbageCount++
			}
		case ">":
			garbage = false
		case "!":
			s.Scan() // Skip next character
		default:
			if garbage {
				garbageCount++
			}
		}
	}
	fmt.Println(groupSum)
	fmt.Println(garbageCount)
}

File: 2017/09/09.py
with open('input', 'r') as f:
    group_level = 0
    group_sum = 0
    garbage = False
    garbage_count = 0

    while True:
        c = f.read(1)
        if not c:
            break

        if c == '{':
            if not garbage:
                group_level += 1
                group_sum += group_level
            else:
                garbage_count += 1
        elif c == '}':
            if not garbage:
                group_level -= 1
            else:
                garbage_count += 1
        elif c == '<':
            if not garbage:
                garbage = True
            else:
                garbage_count += 1
        elif c == '>':
            garbage = False
        elif c == '!':
            f.read(1)
        else:
            if garbage:
                garbage_count += 1

    print(group_sum)
    print(garbage_count)